File: lintcode/binary_tree_zigzag_level_order_traversal.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

from collections import deque
class Solution:
    """
    @param root: The root of binary tree.
    @return: A list of list of integer include 
             the zig zag level order traversal of its nodes' values
    """
    def zigzagLevelOrder(self, root):
        if not root: return []
        queue, result, level = deque([root, None] if root else []), [deque()], 0
        ops = [deque.append, deque.appendleft]
        while queue:
            node = queue.popleft()
            if not node:
                level += 1
                if queue:
                    result.append(deque())
                    queue.append(None)
            else:
                ops[level % 2](result[level], node.val)
                if node.left: queue.append(node.left)
                if node.right: queue.append(node.right)
        return [list(item) for item in result]

File: leetcode/binary_tree_zigzag_level_order_traversal.js
/* Given a binary tree, return the zigzag level order traversal of its nodes' 
 * values. (ie, from left to right, then right to left for the next level and 
 * alternate between).
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var zigzagLevelOrder = function(root) {
    function helper(root, level, levels) {
        if (root) {
            if (levels.length <= level)
                levels.push([]);
            levels[level].push(root.val);
            helper(root.left, level + 1, levels);
            helper(root.right, level + 1, levels);
        }
        return levels;
    }
    var result = helper(root, 0, []);
    for (var i = 1; i < result.length; i += 2)
        result[i].reverse();
    return result;
};
